SDP-50 read event adoption

evtimo opened the pull request.

evtimo committed:
submitEventReadOperation

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

evtimo committed:
revert unused imports

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		@Cleanup val readerGroupManager = ReaderGroupManager.withScope(scopeName, controllerURI);
+		readerGroupManager.createReaderGroup(readerGroup, readerGroupConfig);
+
+		@Cleanup val clientFactory = ClientFactory.withScope(scopeName, controllerURI);
```
The client factory should be obtained from the cache

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		val readerGroupConfig = ReaderGroupConfig.builder()
+				.stream(Stream.of(scopeName, streamName))
+				.build();
+		@Cleanup val readerGroupManager = ReaderGroupManager.withScope(scopeName, controllerURI);
```
this should be cached in the same way as the client factories

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		readerGroupManager.createReaderGroup(readerGroup, readerGroupConfig);
+
+		@Cleanup val clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+		@Cleanup val reader = clientFactory.createReader("reader",
```
should be cached

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		@Cleanup val clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+		@Cleanup val reader = clientFactory.createReader("reader",
+				readerGroup,
+				new JavaSerializer<String>(),
```
ByteBuffer serializer should be used

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		EventRead<String> event = null;
+		try {
+			for (event = reader.readNextEvent(readTimeoutMillis); null != event.getEvent(); ) {
+				if (event.getEvent() != null) {
```
the condition is already passed in the loop statement
2nd invocation will have no effect

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		try {
+			for (event = reader.readNextEvent(readTimeoutMillis); null != event.getEvent(); ) {
+				if (event.getEvent() != null) {
+					System.out.format("Read event '%s'%n", event.getEvent());
```
Let's replace this w/ `Loggers.MSG.trace(...)` call

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -29,33 +31,24 @@
-import io.pravega.client.stream.Serializer;
-import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.client.admin.ReaderGroupManager;
+import io.pravega.client.stream.*;
```
we should not use the wildcard import statements, please fix it

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -64,16 +57,17 @@
+
 public class PravegaStorageDriver<I extends Item, O extends Operation<I>>
-extends CoopStorageDriverBase<I, O>  {
+		extends CoopStorageDriverBase<I, O> {
```
Please rollback all autoformatting related changes. These changes pollute the diff

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
-		// TODO: Evgeny, issue SDP-50
+		val controllerURI = URI.create(uriSchema);
+		val path = evtOp.dstPath();
+		val scopeName = path.substring(0, path.indexOf("/"));
```
If I remember correctly, we don't store scope in the operation anymore

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -386,63 +380,90 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		} catch (ReinitializationRequiredException e) {
+			e.printStackTrace();
+		}
+		System.out.format("No more events from %s/%s%n", scopeName, streamName);
```
"Let's replace this w/ Loggers.MSG.trace(...) call" as Andrey says. I think it should be applied here as well

evtimo committed:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-pravega into issue-50-readEventAdoption

Signed-off-by: Evgeny <zheny.tim@gmail.com>

# Conflicts:
#	src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java

evtimo committed:
cache added + fixes

Signed-off-by: Evgeny <zheny.tim@gmail.com>

evtimo committed:
import fixes

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

evtimo committed:
import fixes

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -172,6 +184,12 @@ public final StreamConfiguration apply(final String streamName) {
+	private final Map<URI, ReaderGroupManagerCreateFunction> readerGroupManagerCreateFunctionMapCache = new ConcurrentHashMap<>();
+	private final Map<String, ReaderGroupManager> readerGroupManagerCache = new ConcurrentHashMap<>();
+	// * event stream reader
+	private final Map<ClientFactory, ReaderCreateFunction> eventStreamReaderCreateFunctionMapCache = new ConcurrentHashMap<>();
```
the word map seems excessive

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -431,7 +449,39 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		val endpointUri = endpointCache.computeIfAbsent(nodeAddr, this::createEndpointUri);
+		val path = evtOp.dstPath();
+		val scopeName = DEFAULT_SCOPE;
+		val streamName = path.substring(path.indexOf("/") + 1);
```
Andrey wrote a nice peace of code in eventCreate method on extracting a right streamName

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -431,7 +449,39 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+			}
+		} catch (ReinitializationRequiredException e) {
+			e.printStackTrace();
+		}
```
I think you forgot to call all the completion methods. Look at eventCreate for an example

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/cache/ReaderCreateFunctionImpl.java:
```diff
@@ -0,0 +1,20 @@
+
+	@Override
+	public EventStreamReader apply(String readerGroup) {
+		return clientFactory.createReader("reader", readerGroup, new DataItemSerializer(true), ReaderConfig.builder().build());
```
Dont we have an DataItemSerializer as a class field? evtSerializer i guess?

akurilov replied:
I guess we should use some kind of the ByteBuffer [de]serializer. Because the DataItemSerializer cannot deserialize. Also, we have a working example in the tests.

evtimo committed:
fixes after review

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

dadlex commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -431,7 +449,39 @@ void submitEventCreateOperation(final DataOperation evtOp, final String nodeAddr
+		Loggers.MSG.trace("Reading all the events from {} {}", scopeName, streamName);
+		EventRead<String> event = null;
+		try {
+			for (event = reader.readNextEvent(readTimeoutMillis); null != event.getEvent(); ) {
```
I'm totally lost now. @akurilov  For now the only difference between eventRead and streamRead is how we get the metrics? So the difference will be in completeEventRead vs completeStreamRead?

akurilov replied:
Probably yes but I'm not sure

evtimo committed:
issue 30 uriSchema configurable

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

evtimo committed:
issue 30 uriSchema configurable

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

evtimo committed:
Merge remote-tracking branch 'origin/issue-50-readEventAdoption' into issue-50-readEventAdoption

evtimo committed:
fix

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>
